dozer-recordstore/src/lib.rs

```diff
@@ -0,0 +1,542 @@
+//! [`RecordRef`] is a compact representation of a collection of [dozer_types::types::Field]s
+//! There are two principles that make this representation more compact than `[Field]`:
+//!  1. The fields and their types are stored as a Struct of Arrays instead of
+//!     and Array of Structs. This makes it possible to pack the discriminants
+//!     for the field types as a byte per field, instead of taking up a full word,
+//!     which is the case in [Field] (because the variant value must be aligned)
+//!  2. The field values are stored packed. In a `[Field]` representation, each
+//!     field takes as much space as the largest enum variant in [Field] (plus its discriminant,
+//!     see (1.)). Instead, for the compact representation, we pack the values into
+//!     align_of::<Field>() sized slots. This way, a u64 takes only 8 bytes, whereas
+//!     a u128 can still use its 16 bytes.
+use std::alloc::{dealloc, handle_alloc_error, Layout};
+use std::{hash::Hash, ptr::NonNull};
+
+use triomphe::{Arc, HeaderSlice};
+
+use dozer_types::chrono::{DateTime, FixedOffset, NaiveDate};
+use dozer_types::json_types::JsonValue;
+use dozer_types::ordered_float::OrderedFloat;
+use dozer_types::rust_decimal::Decimal;
+use dozer_types::types::{DozerDuration, DozerPoint};
+use dozer_types::{
+    serde::{Deserialize, Serialize},
+    types::{Field, FieldType, Lifetime},
+};
+
+// The alignment of an enum is necessarily the maximum alignment of its variants
+// (otherwise it would be unsound to read from it).
+// So, by using the alignment of `Field` as the alignment of the values in our
+// packed `RecordRef`, we ensure that all accesses are aligned.
+// This wastes a little bit of memory for subsequent fields that have
+// smaller minimum alignment and size (such as `bool`, which has size=1, align=1),
+// but in practice this should be negligible compared to the added effort of
+// packing these fields while keeping everything aligned.
+const ALIGN: usize = std::mem::align_of::<Field>();
+
+// This asserts at compile time that `ALIGN` is not
+// accidentally increased, as that would waste space.
+const _ASSERT_ALIGN: usize = ALIGN - 8;
```

Comment:
Whoops, should have been the other way around so that it underflows when ALIGN > 8

Reply:
OK, it doesn't compile on my M1 becuase my `ALIGN` is 16.

Reply:
See https://github.com/rust-lang/rust/issues/54341

Reply:
Hmm. If we align everything to 16 bytes, the memory savings become quite small. What I can do is add an alignment addition to the start of `read` and `write`, so that the alignment of fields is dynamic according to their types.

Reply:
We can merge this because it's a decent save on x86_64

Reply:
I actually got it working surprisingly quickly. I'll push after my fuzzing run is sufficiently long. After that would you mind fuzzing it as well to make sure it's also sound on apple silicon?

Reply:
Sure I'll do that

Reply:
The alignment is not a soundness issue so we're probably good here

Reply:
Creating an unaligned reference is immediate UB, even without reading it, so I'd like to make sure. Same with unaligned pointer reads.

Reply:
Point is the alignment **is** 8 on x86_64, otherwise it's a Rust soundness bug.